arelle/plugin/validate/NL/rules/br_kvk.py

```diff
@@ -0,0 +1,251 @@
+"""
+See COPYRIGHT.md for copyright information.
+"""
+from __future__ import annotations
+
+from datetime import date, datetime, timedelta
+from typing import Any, Iterable, TYPE_CHECKING
+
+from arelle.ValidateXbrl import ValidateXbrl
+from arelle.typing import TypeGetText
+from arelle.utils.PluginHooks import ValidationHook
+from arelle.utils.validate.Decorator import validation
+from arelle.utils.validate.Validation import Validation
+from . import (
+    DOCUMENTATION_ADOPTION_DATE_QN,
+    DOCUMENTATION_ADOPTION_STATUS_QN,
+    DOCUMENT_RESUBMISSION_UNSURMOUNTABLE_INACCURACIES_QN,
+    FINANCIAL_REPORTING_PERIOD_CURRENT_END_DATE_QN,
+    FINANCIAL_REPORTING_PERIOD_CURRENT_START_DATE_QN,
+    FINANCIAL_REPORTING_PERIOD_PREVIOUS_END_DATE_QN,
+    FINANCIAL_REPORTING_PERIOD_PREVIOUS_START_DATE_QN,
+)
+from ..DisclosureSystems import (
+    DISCLOSURE_SYSTEM_NT16,
+    DISCLOSURE_SYSTEM_NT17,
+)
+from ..PluginValidationDataExtension import PluginValidationDataExtension
+
+if TYPE_CHECKING:
+    from arelle import ModelXbrl
+    from arelle.ModelValue import QName
+
+_: TypeGetText
+
+
+def _getReportingPeriodDateValue(modelXbrl: ModelXbrl, qname: QName) -> date:
+    facts = modelXbrl.factsByQname.get(qname)
+    assert facts and len(facts) == 1, _('Exactly one fact must exist for reporting period concept: {0}').format(qname)
```

Comment:
It seems wrong to raise an exception if there are no facts, if there are duplicate facts (I assume there's a separate rule that intends to diagnose that, or if not, we should somehow warn once if we find duplicate period facts?), or the fact has an invalid value.

Reply:
There are separate checks for making sure there are exactly one of most of concepts involved in these validations. Would the best practice be to have these validations fail silently and let those other checks draw attention to any missing/duplicate fact issues? Or should an issue like this that prevents the actual logic of the validation from running trigger the failure of that validation in the same way?

Reply:
> There are separate checks for making sure there are exactly one of most of concepts involved in these validations.

Which ones don't have a "missing" separate check?  If it's previous period (e.g., new company?), perhaps we just need to tolerate the facts as missing (i.e., current and current-1day rather than current,previous,previous-1day).  Skipping this validation for missing mandatory facts seems reasonable.

Which ones don't have a separate "duplicate" check?  I'm not sure what to do since XBRL doesn't forbid duplicate facts ([notes](https://www.xbrl.org/WGN/xbrl-duplicates/WGN-2015-12-09/xbrl-duplicates-WGN-2015-12-09.html#sec-approaches)), so unless their manual does, it's unclear what the regulator enforces (no duplicates anywhere, no duplicates for these facts, or allow _all_ values as valid).  Perhaps start by adding duplicate checking for the key facts in this rule, report them with this rule's error code, and add a comment explaining we're not sure which behavior is actually intended.

If any single key fact value is invalid, I'd say skip the validation.  There's no use in reporting an issue for every other fact in the report saying they have invalid context just because some key fact has a typo.

Reply:
Agreed with Brett. Further, these validation rules should never raise an exception. They should either pass or fail (by returning one or more validation warnings/errors).

Reply:
It might be worthwhile to have a certain exception type (could be `AssertionError`, could be something specific to this validation plugin design) that can be thrown from validations and then be logged as a debug message and ignored. 

That way rather than having to, for example, modify `_getReportingPeriodDateValue` to potentially return `None`, then handle the `None` case to gracefully exit the validation, we can just immediately `raise IncompleteValidationError('Exactly one fact must exist for...')` once we know the validation won't be possible to check. I suppose this only really matters if we often make use of helper functions, otherwise it's no better than having a `return` statement. Thoughts?

Reply:
Seems plausible.  It might be nice if the exception contained a code+message so that we could fail harder if the code was not produced elsewhere/later (even if just as part of testing; e.g., a new plugin callback for handling IncompleteValidationError so the test harness can collect them until the end).